core/pta/attestation.c

```diff
@@ -0,0 +1,176 @@
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct vm_region *r = NULL;
+	struct vm_region **regions = NULL;
+	size_t nregions = 0;
+	void *ctx = NULL;
+	size_t i = 0;
+
+	res = crypto_hash_alloc_ctx(&ctx, TEE_ALG_SHA256);
+	if (res)
+		return res;
+
+	res = crypto_hash_init(ctx);
+	if (res)
+		goto out;
+
+	/*
+	 * Make an array of region pointers so we can use qsort() to order it.
+	 */
+
+	TAILQ_FOREACH(r, &vm_info->regions, link)
+		if (is_region_valid(r))
+			nregions++;
+
+	regions = malloc(nregions * sizeof(*regions));
+	if (!regions) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	TAILQ_FOREACH(r, &vm_info->regions, link)
+		if (is_region_valid(r))
+			regions[i++] = r;
+
+	/*
+	 * Sort regions so that they are in a consistent order even when TA ASLR
+	 * is enabled.
+	 */
+	qsort(regions, nregions, sizeof(*regions), cmp_regions);
+
+	/* Hash regions in order */
+	for (i = 0; i < nregions; i++) {
+		r = regions[i];
+		DMSG("va %p size %zu", (void *)r->va, r->size);
+		res = crypto_hash_update(ctx, (uint8_t *)r->va, r->size);
+		if (res)
+			goto out;
+	}
+
+	res = crypto_hash_final(ctx, hash, TEE_SHA256_HASH_SIZE);
+out:
+	free(regions);
+	crypto_hash_free_ctx(ctx);
+	return res;
+}
+
+static TEE_Result hash_ta(uint32_t types, TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_session *s = NULL;
+	struct ts_session *ts_sess = NULL;
+	uint32_t id = params[0].value.a;
+	uint8_t *hash = params[1].memref.buffer;
+	size_t hash_sz = params[1].memref.size;
+	TEE_Result ret = TEE_SUCCESS;
+	struct user_ta_ctx *ctx = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+				     TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE, TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!hash || hash_sz != TEE_SHA256_HASH_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (ts_get_calling_session()) {
+		/*
+		 * Called from secure world: not what we expect. Only the normal
+		 * world CA is supposed to know about the session ID.
+		 */
+		return TEE_ERROR_ACCESS_DENIED;
+	}
+
+	nsec_sessions_list_head(&open_sessions);
+	s = tee_ta_get_session(id, true, open_sessions);
+	if (!s)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (!is_user_ta_ctx(s->ts_sess.ctx))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	ts_push_current_session(&s->ts_sess);
+
+	ctx = to_user_ta_ctx(s->ts_sess.ctx);
+	ret = hash_regions(&ctx->uctx.vm_info, hash);
+
+	ts_sess = ts_pop_current_session();
+	assert(ts_sess == &s->ts_sess);
+
+	tee_ta_put_session(s);
+	return ret;
+}
+
+static TEE_Result invoke_command(void *pSessionContext __unused,
+				 uint32_t nCommandID, uint32_t nParamTypes,
+				 TEE_Param pParams[TEE_NUM_PARAMS])
+{
+	switch (nCommandID) {
+	case PTA_ATTESTATION_HASH_TA:
+		return hash_ta(nParamTypes, pParams);
+	default:
+		break;
+	}
+
+	return TEE_ERROR_NOT_IMPLEMENTED;
+}
+
+pseudo_ta_register(.uuid = PTA_ATTESTATION_UUID, .name = PTA_NAME,
+		   .flags = PTA_DEFAULT_FLAGS,
```

Comment:
Likely some feature that deserves enumeration.

Reply:
In my mind the caller is typically a user space CA, not the REE kernel so that might not be very useful.